packages/action/index.js

```diff
@@ -22,32 +22,78 @@ const core = require('@actions/core');
 async function main() {
   try {
     // `who-to-greet` input defined in action metadata file
+    if(!core.getInput('repo-token') || !core.getInput('os') || !core.getInput('file-path') || !core.getInput('repo-token') ){
+      throw "Github action missing required fields. Refer to documentation at https://github.com/googlecloudplatform/flaky-service";
+    }
+
+   
+    let buildmessageStr = process.env.GITHUB_WORKFLOW + ' ' + process.env.GITHUB_RUN_NUMBER;
+    if(core.getInput('tag') && core.getInput('tag') !== "None"){
+      buildmessageStr += ' (' + core.getInput('tag') + ')';
+    }
+
 
     const metaData = {
-      github: JSON.parse(core.getInput('github')),
-      os: JSON.parse(core.getInput('os')),
-      matrix: JSON.parse(core.getInput('matrix')),
+      repoId: encodeURIComponent(process.env.GITHUB_REPOSITORY), //resanitized server side
+      organization: process.env.GITHUB_REPOSITORY.substr(0,process.env.GITHUB_REPOSITORY.indexOf("/")),
+      timestamp: new Date(),
+      url: process.env.GITHUB_SERVER_URL + '/' + process.env.GITHUB_REPOSITORY,
+      environment:{
+        os: core.getInput('os'),
+        tag: core.getInput('tag'),
+        matrix :JSON.parse(core.getInput('matrix')),
+        ref: process.env.GITHUB_REF
+      },
+      buildId: process.env.GITHUB_RUN_ID,
+      sha: process.env.GITHUB_SHA,
+      name: process.env.GITHUB_REPOSITORY.substr(1+process.env.GITHUB_REPOSITORY.indexOf("/")),
+      description: core.getInput('repo-description'),
+      buildmessage: buildmessageStr,
+      token: core.getInput('repo-token')
     };
-    const fileType = core.getInput('logtype');
-    console.log('reading from: ' + core.getInput('filepath'));
-    const data = fs.readFileSync(
-        core.getInput('filepath'), 'utf8');
 
+    metaData.matrix = JSON.stringify(metaData.environment.matrix, Object.keys(metaData.environment.matrix).sort()); //consistancy 
```

Comment:
curious what the reasoning is for sending this to the API as a serialized string, vs., just leaving it an array. What's the ultimate representation in the database?

Reply:
In the database, its stored as a string. It keys need to be in alphabetic order because  in order to have a dropdown which filters by filter, you need to check for equality between the matrices. This could be done differently though. Another motivation for storing it as a string is to avoid a user input to be a Firestore key.